File: 0053-maximum-subarray/0053-maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        int m = INT_MIN;
        int flag = 0;
        int ssum=0,msum=0;
        for(int i=0;i<n;i++){
            if(nums[i]<0){
                m = max(m,nums[i]);
            }
            if(nums[i] == 0){
                flag = 1;
            }
            ssum += nums[i];
            if(ssum>msum){
                msum = ssum;
            }
            if(ssum<0){
                ssum = 0;
            }
        }
        if(msum == 0 and flag==0){
            return m;
        }
        return msum;
    }
};

File: 0053-maximum-subarray/0053-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    let hasPositive = false, currMax = 0, max = 0, maxNeg = nums[0];
    for(let i=0;i< nums.length;i++){
        if(nums[i] > 0){
            hasPositive = true;
        }
        else{
            maxNeg = Math.max(maxNeg, nums[i]);
        }
        currMax += nums[i];
        max = Math.max(currMax, max);
        if(currMax < 0){
            currMax = 0;
        }
    }
    if(hasPositive !== true){
        return maxNeg;
    }
    return max;
};